packages/NodeCollector/BinaryOpTreeRootLocator.php

```diff
@@ -20,17 +23,22 @@ final class BinaryOpTreeRootLocator
      */
     public function findOperationRoot(Expr $expr, string $binaryOpClass): Expr
     {
-        /** @var ?Expr $parentNode */
         $parentNode = $expr->getAttribute(AttributeKey::PARENT_NODE);
-        if ($parentNode === null || $parentNode::class !== $binaryOpClass) {
+        if (! $parentNode instanceof Expr) {
```

Comment:
Interesting, how does it know that this transformation is valid?

While technically the only value of `$parentNode` that matches `! $parentNode instanceof Expr` is `null`, this change does not feel like it adds clarity to the code, rather the opposite.

Reply:
Based on `/** @var ?Expr $parentNode */`, which is incorrect.

The check should be to `! parentNode instanceof Node`

Reply:
I updated to use instanceof Node check https://github.com/rectorphp/rector-src/pull/1621/commits/fd2ce4bb7a47f59f5002be743f124b8b7de5e53b

Reply:
> Based on /** @var ?Expr $parentNode */, which is incorrect.

Yeah, I suppose. But I think the condition should still be `$parentNode === null` – we do not really care whether the parent node is subclass of `Node`, only if `$parentNode::class !== $binaryOpClass`, so the `instanceof` check would only be a less clear synonym of `null` check.

Reply:
The `$node instanceof Node` keeps the type, that's why.